From 90c3c906a38767362d9fef59f57d754c8de5e1bd Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Wed, 24 Jul 2024 18:23:22 +0200
Subject: [PATCH 1050/1695] drivers: stm32_risab: support non-TDCID probe and
 RIF config application

In case we're booting as non-TDCID, the RISAB1/2 are still only
configurable by the cortexA. For others, keep the same configuration
sequence.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 375b6801792747825d99d3c780d343873ae317b8 drivers: stm32_risab: support non-TDCID probe and RIF config application]
Change-Id: I3d6f8a4e7f63056a78c9f647f9f020e427e12525
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/395782
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/platform_config.h |  2 ++
 core/drivers/firewall/stm32_risab.c           | 21 ++++++++++++-------
 2 files changed, 15 insertions(+), 8 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/platform_config.h b/core/arch/arm/plat-stm32mp2/platform_config.h
index 72a0676e9..949f93883 100644
--- a/core/arch/arm/plat-stm32mp2/platform_config.h
+++ b/core/arch/arm/plat-stm32mp2/platform_config.h
@@ -56,6 +56,8 @@
 #define RIFSC_BASE			0x42080000
 #define RISAF4_BASE			0x420d0000
 #define RISAF5_BASE			0x420e0000
+#define RISAB1_BASE			0x420f0000
+#define RISAB2_BASE			0x42100000
 #define RISAB6_BASE			0x42140000
 #define BSEC3_BASE			0x44000000
 #define IWDG2_BASE			0x44002000
diff --git a/core/drivers/firewall/stm32_risab.c b/core/drivers/firewall/stm32_risab.c
index 21049f010..4be5a9f8a 100644
--- a/core/drivers/firewall/stm32_risab.c
+++ b/core/drivers/firewall/stm32_risab.c
@@ -304,15 +304,20 @@ static void apply_rif_config(struct stm32_risab_pdata *risab_d)
 	for (i = 0; i < risab_d->nb_regions_cfged; i++) {
 		set_block_dprivcfgr(risab_d, &risab_d->subr_cfg[i]);
 
-		/* Delegate RIF configuration or not */
-		if (!is_tdcid)
-			DMSG("Cannot set %s CID configuration for region %u",
-			     risab_d->risab_name, i);
-		else
+		if (virt_to_phys((void *)base) != RISAB1_BASE &&
+		    virt_to_phys((void *)base) != RISAB2_BASE) {
+			/* Delegate RIF configuration or not */
+			if (!is_tdcid)
+				DMSG("Cannot set %s CID config for region %u",
+				     risab_d->risab_name, i);
+			else
+				set_cidcfgr(risab_d, &risab_d->subr_cfg[i]);
+
+			if (!regs_access_granted(risab_d, i))
+				panic();
+		} else {
 			set_cidcfgr(risab_d, &risab_d->subr_cfg[i]);
-
-		if (!regs_access_granted(risab_d, i))
-			panic();
+		}
 
 		/*
 		 * This sequence will generate an IAC if the CID filtering
-- 
2.47.2

